### ISSUE TITLE
How do I fix this error "Could not commit changes: copy failed: fl:filesys​tem:System​Error:" please?

### ISSUE THREAD

**Issue by cuixing158** [mathworks-ref-arch/matlab-dockerfile#61]
When I try to compile and generate a simple typical mex(or lib,dll,exe) function on an ubuntu system, I get the following error, when the mex file can be generated and run normally, but I don't know what causes the error,the report didn't come out either ,how can i fix it? (Run R2022b, ubuntu20.04，built with dockerfile, by `-broswer` flag)

myadd.m
```matlab
function c = myadd(a,b)%#codegen
c = a+b;
end
```
command line:
```matlab
a=1;
b=2;
codegen -config:mex myadd -args {a,b} -report
```
Could not commit changes:  copy failed: fl:filesystem:SystemError:
/tmp/.fa30-a34f-2ae0-0e8e.tmp Invalid cross-device link 

---
Similar problems occur when saving mlx files, but these can be solved by setting the "TEMPDIR" environment variable in the desktop version of matlab, but not in matlab containers built with dockerfile.

**Comment by prabhakk-mw** [comment]
@cuixing158 Can you please share your Dockerfile?

**Comment by prabhakk-mw** [comment]
The filesystem permission issues are mostly due to the incorrect USER running inside container. In the Dockerfile shared above you will be running as ROOT user.

The suggested repro steps work for me after, I've made a couple of minor updates to the Dockerfile:

1. Added a `||` operator after the `MATLAB_Coder \` line. Without which the Dockerfile would not build successfully.
2. Added a `-f` flag to the  `ln` command.
3. Switched the user to the default `matlab` user at the end of the Dockerfile
4. Switched the WORKDIR to a location that the `matlab` user has permissions to run in.

See below:
```Dockerfile
ARG MATLAB_DOCKER_TAG=latest
FROM mathworks/matlab:${MATLAB_DOCKER_TAG}
USER root
WORKDIR /root
RUN export DEBIAN_FRONTEND=noninteractive && apt-get update && \
    apt-get install --no-install-recommends --yes wget \
    build-essential \ 
    cmake && \
    apt-get clean && apt-get -y autoremove && rm -rf /var/lib/apt/lists/*

WORKDIR /opt/matlab
RUN wget -q https://www.mathworks.com/mpm/glnxa64/mpm && \
    chmod +x mpm && \
    MATLAB_RELEASE=`ls | grep R*` && \
    ./mpm install --destination=/opt/matlab/${MATLAB_RELEASE}/ --release=${MATLAB_RELEASE} \
    MATLAB_Coder \
    || (echo "MPM Installation Failure. See below for more information:" && cat /tmp/mathworks_root.log && false) && \
    rm -f mpm /tmp/mathworks_root.log && \
    ln -fs /opt/matlab/${MATLAB_RELEASE}/bin/matlab /usr/local/bin/matlab

USER matlab
WORKDIR /home/matlab
```

Finally I used the following commands to build and run the Dockerfile:
```bash
$ docker build -t matlab-dockerfile:issue-61-localuser -f Dockerfile . 
$ docker run -it --rm -p 8888:8888 matlab-dockerfile:issue-61-localuser -browser
```
Below is the screen shot showing it working:
![chrome-capture-2023-2-8](https://user-images.githubusercontent.com/64955767/223700008-bd22906f-282a-43ec-8ba4-6ba477e1f8d3.png)

**Comment by cuixing158** [comment]
@prabhakk-mw 
I recently ran into a slightly more troublesome issue where above docker matlab container build's mex file dependency on the opencv c++ library would cause the dependency to not be found, see the [link here](https://www.mathworks.com/matlabcentral/answers/1754390-matlab-crashing-in-ubuntu-when-using-matlab-coder#comment_2652515) for details. I've tried hard to write simple dependencies on opencv c++ for mex file generation, but it never worked.
BTW,If you need to test sample c++ and matlab code, I can share it with you

**Comment by prabhakk-mw** [comment]
@cuixing158, its probably a good idea to contact [Mathworks Technical Support](https://mathworks.com/help/matlab/matlab_env/contact-technical-support.html) and create a service request to help you resolve this issue.

And sure, please share the steps to reproduce this at my end, to see if there is anything that I can help troubleshoot this with.

**Comment by cuixing158** [comment]
Hi,@prabhakk-mw,
The problem can finally be solved, thanks to the meticulous and professional investigation by TMW staff, this docker matlab mex opencv error has now been successfully reolved.Not related to matlab coder toolbox.

Hi,
Steps to reproduce the error (all steps are performed in the docker container).

1. The installation was successfully compiled from the opencv4.6.0 source code in matlab docker, and "pkg-config --modversion opencv4" correctly returned the version of opencv as 4.6.0, plus the separate opencv program compiled and ran correctly under gcc/g++, a step I am sure is fine.
2. set the "LD_LIBRARY_PATH" environment variable for OPENCV with the "setenv" function, as "setenv("LD_LIBRARY_PATH", ['/home/matlab/opencv_4_6_0/lib:',getenv("LD_LIBRARY_PATH")])".
3. I wrote the simple entry function "imread_opencv.m" and used the "coder.ExternalDependency" class to call the dependent opencv program in the matlab environment.
4.  successfully generated the "imread_opencv_mex.mexa64" file for the entry function via the command line "codegen".
5.  Running the test image "test_480x640_uint8.jpg", it reports the error that I gave above.
[image: bug_reproduce.jpg]

---
I set "**LD_LIBRARY_PATH**" within matlab-container before generate mex file , below is return value . I have tried to set the "LD_LIBRARY_PATH" path to the opencv library in the container system (~/.bashrc) several times, but the bug still occurs both before and after mex.
```matlab
getenv("LD_LIBRARY_PATH")
```
```text
ans =


'/home/matlab/opencv_4_6_0/lib:/opt/matlab/R2022b/sys/opengl/lib/glnxa64:/opt/matlab/R2022b/sys/os/glnxa64:/opt/matlab/R2022b/bin/glnxa64:/opt/matlab/R2022b/extern/lib/glnxa64:/opt/matlab/R2022b/cefclient/sys/os/glnxa64:/opt/matlab/R2022b/sys/java/jre/glnxa64/jre/lib/amd64/native_threads:/opt/matlab/R2022b/sys/java/jre/glnxa64/jre/lib/amd64/server'
```
---
The problem was that the environment variable **LD_LIBRARY_PATH** was not effectively recognised by the docker matlab container. The current version of the R2022b container, once built, is not affected by the environment variables set in the container (i.e. neither setting **LD_LIBRARY_PAH** on the system nor using the setenv function within matlab can actually change its value within the container, even though getenv can ostensibly "show the results in effect").

To properly set environment variables to be passed to matlab, they should be passed directly when matlab is started, e.g. started in the following way.
```bash
LD_LIBRARY_PATH="REPLACE/THIS/PATH" matlab
```
or run matlab container
```bash
docker run -d --rm -p <host-port-id>:<contain-port-id> --shm-size=512M -e
LD_LIBRARY_PATH=<my_opencv_lib_dir>  <my_image_name>:r2022b -browser
```
\<host-port-id>:Replace with yourself
\<container-port-id>:Replace with yourself
\<my_opencv_lib_dir>  : Replace with your own path
\<my_image_name> : Replace with your own name
